File: leetcode/q6.java
package leetcode;
import java.util.*;

public class q6 {
    class Solution {
        public String convert(String s, int numRows) {
            if(numRows == 1)  return s;
    
            int cycle = numRows*2 -2;
    
            List<StringBuilder> sbs = new ArrayList<>();
            for(int i = 0; i < numRows; i++) {
                sbs.add(new StringBuilder());
            }
    
            int index = 0;
            while(index < s.length()) {
                int remainder = index%cycle;
                if(remainder >= numRows) remainder = cycle-remainder;
    
                sbs.get(remainder).append(s.charAt(index++));
            }
    
            for(int i = 1; i < numRows; i++) {
                sbs.get(0).append(sbs.get(i));
            }
    
            return sbs.get(0).toString();
        }
    }
}

File: top150/_01_ArrayString/q6.py
class Solution:
    # T: O(N), S: O(M*N); N: len(s), M: numRows
    def convert(self, s: str, numRows: int) -> str:
        if numRows == 1 or len(s) < numRows:
            return s

        rows = [''] * numRows
        n = len(s)
        r = 0
        order = -1

        for ch in s:
            rows[r] += ch
            if r == 0 or r == numRows-1:
                order *= (-1)
            if order == 1:
                r += 1
            elif order == -1:
                r -= 1
        return ''.join(rows)
